File: src/main/java/cn/labzen/tool/util/Systems.java
package cn.labzen.tool.util;

import java.io.File;
import java.nio.file.FileSystems;

public final class Systems {

  private static final OS CUR_OS;

  static {
    String osName = System.getProperty("os.name").toLowerCase();
    if (osName.contains("win")) {
      CUR_OS = OS.WINDOWS;
    } else if (osName.contains("nix") || osName.contains("nux") || osName.contains("aix")) {
      boolean android = false;
      try {
        Class.forName("android.os.Build");
        android = true;
      } catch (ClassNotFoundException e) {
        // do nothing
      }
      CUR_OS = android ? OS.ANDROID : OS.LINUX;
    } else if (osName.contains("mac")) {
      CUR_OS = OS.MAC;
    } else if (osName.contains("sunos")) {
      CUR_OS = OS.SOLARIS;
    } else {
      CUR_OS = OS.UNKNOWN;
    }
  }

  private Systems() {
  }

  public static OS getOS() {
    return CUR_OS;
  }

  public static int getOSArch() {
    return System.getProperty("os.arch").contains("64") ? 64 : 32;
  }

  public static String javaVersion() {
    return System.getProperty("java.version");
  }

  public static String javaVendor() {
    return System.getProperty("java.vendor");
  }

  public static String javaHome() {
    return System.getProperty("java.home");
  }

  public static String javaClasspath() {
    return System.getProperty("java.class.path");
  }

  public static String osVersion() {
    return System.getProperty("os.version");
  }

  public static String fileSeparator() {
    return FileSystems.getDefault().getSeparator();
  }

  public static String pathSeparator() {
    return File.pathSeparator;
  }

  public static String lineSeparator() {
    return System.lineSeparator();
  }

  public static String userName() {
    return System.getProperty("user.name");
  }

  public static String userHome() {
    return System.getProperty("user.home");
  }

  public static String userDir() {
    return System.getProperty("user.dir");
  }

  public enum OS {
    WINDOWS, LINUX, MAC, SOLARIS, ANDROID, UNKNOWN
  }
}

File: src/main/kotlin/cn/labzen/tool/util/Systems.kt
package cn.labzen.tool.util

import java.io.File
import java.nio.file.FileSystems
import java.util.*

object Systems {

  /**
   * 系统名称
   */
  @JvmStatic
  val osName: OS by lazy {
    val name = System.getProperty("os.name").lowercase(Locale.getDefault())

    when {
      name.contains("win") -> OS.WINDOWS
      name.contains("nix") || name.contains("nux") || name.contains("aix") ->
        try {
          Class.forName("android.os.Build")
          OS.ANDROID
        } catch (e: ClassNotFoundException) {
          OS.LINUX
        }

      name.contains("mac") -> OS.MAC
      name.contains("sunos") -> OS.SOLARIS
      else -> OS.UNKNOWN
    }
  }

  /**
   * JVM是多少位系统，返回只有32和64两个值
   */
  @JvmStatic
  val osArch: Int by lazy {
    val arch = System.getProperty("os.arch")
    if (arch.contains("64")) 64 else 32
  }


  // system properties
  @JvmField
  val javaVersion: String = System.getProperty("java.version")

  @JvmField
  val javaVendor: String = System.getProperty("java.vendor")

  @JvmField
  val javaHome: String = System.getProperty("java.home")

  @JvmField
  val javaClasspath: String = System.getProperty("java.class.path")

  @JvmField
  val osVersion: String = System.getProperty("os.version")

  @JvmField
  val fileSeparator: String = FileSystems.getDefault().separator

  @JvmField
  val pathSeparator: String = File.pathSeparator

  @JvmField
  val lineSeparator: String = System.lineSeparator()

  @JvmField
  val userName: String = System.getProperty("user.name")

  @JvmField
  val userHome: String = System.getProperty("user.home")

  @JvmField
  val userDir: String = System.getProperty("user.dir")

  enum class OS {
    WINDOWS, LINUX, MAC, SOLARIS, ANDROID, UNKNOWN
  }
}
